tools/ci_connector_ops/ci_connector_ops/pipelines/publish.py

```diff
@@ -0,0 +1,232 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import json
+import os
+import uuid
+from abc import ABC
+from typing import List, Tuple
+
+import anyio
+import dagger
+from ci_connector_ops.pipelines import builds
+from ci_connector_ops.pipelines.actions import environments
+from ci_connector_ops.pipelines.actions.remote_storage import upload_to_gcs
+from ci_connector_ops.pipelines.bases import ConnectorReport, Step, StepResult, StepStatus
+from ci_connector_ops.pipelines.contexts import ConnectorContext
+from ci_connector_ops.pipelines.pipelines import metadata
+from ci_connector_ops.pipelines.utils import with_stderr, with_stdout
+from dagger import Container, File, QueryError, Secret
+
+
+class PublishStep(Step, ABC):
+    def __init__(self, context: ConnectorContext, pre_release: bool = True) -> None:
+        super().__init__(context)
+        self.pre_release = pre_release
+
+    @property
+    def docker_image_name(self):
+        if self.pre_release:
+            return f"{self.context.docker_image_from_metadata}-dev.{self.context.git_revision[:10]}"
+        else:
+            return self.context.docker_image_from_metadata
+
+
+class CheckConnectorImageDoesNotExist(PublishStep):
```

Comment:
Looks like we have some competing PRs!
https://github.com/airbytehq/airbyte/pull/25306

Reply:
In the very least its not bad to have backup logic :)

Reply:
Interesting! It does not hurt of course to perform this check two times but I think it raises a separation of concerns question.
Should the metadata service be aware of the life cyle of the artifacts it refers to? 
IMO it's the continuous deployment logic that should provide guarantees.  
In the same logic as we're not running connectors test on publish because the fact it's been merged on master with the required tests to pass gives us the guarantee we're publishing an healthy image.
I'd be more inclined to implement a frequent (nightly?) health check on metadata files to be alerted about potential inconsistencies.